src/Hl7.Fhir.Core/Model/DomainResource.cs

```diff
@@ -56,5 +56,32 @@ public override IEnumerable<ValidationResult> Validate(ValidationContext validat
 
             return result;
         }
+
+        /// <summary>
+        /// Finds the contained resource defined by the <paramref name="reference"/>. A reference to a contained resource starts with the
+        /// character #.
+        /// </summary>
+        /// <param name="reference">the reference to contained resource</param>
+        /// <returns>The resource referenced by <paramref name="reference"/>, null otherwise.</returns>
+        public Resource FindContainedResource(ResourceReference reference)
+        {
+            if (!reference.IsContainedReference) return null;
+
+            if (reference.Reference == "#") return this;
+
+            // search the contained resource by removing '#' 
+            return Contained.FirstOrDefault(c => c.Id == reference.Reference.Remove(0, 1));
```

Comment:
This fails if reference.Reference is null.

Reply:
Oops, you are right of course. I created a new commit for that